src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+                            "VAX byte ordering is unsupported for complex number type conversions");
+        }
+
+next:
+        /*
+         * If we had used a temporary buffer for the destination then we
+         * should copy the value to the true destination buffer.
+         */
+        if (d == dbuf) {
+            if (H5T_FLOAT == dst_p->shared->type)
+                H5MM_memcpy(dp, d, dst_p->shared->size);
+            else
+                H5MM_memcpy(dp, d, dst_p->shared->size / 2);
+        }
+
+        /* Ensure imaginary part of complex number is zeroed */
+        if (H5T_COMPLEX == dst_p->shared->type)
+            memset(dp + (dst_p->shared->size / 2), 0, dst_p->shared->size / 2);
+
+        /* Advance source & destination pointers by delta amounts */
+        sp += src_delta;
+        dp += dst_delta;
+    } /* end conversion loop */
+
+done:
+    H5MM_free(src_rev);
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5T__conv_complex_f_matched() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5T__conv_complex_compat
+ *
+ * Purpose:     Performs a no-op conversion between a complex number type
+ *              and an equivalent datatype. Complex number types are
+ *              considered equivalent to the following:
+ *
+ *              - An array datatype consisting of two elements where each
+ *                element is of the same floating-point datatype as the
+ *                complex number type's base floating-point datatype
+ *
+ *              - A compound datatype consisting of two fields where each
+ *                field is of the same floating-point datatype as the
+ *                complex number type's base floating-point datatype. The
+ *                compound datatype must not have any leading or trailing
+ *                structure padding or any padding between its two fields.
+ *                The fields must also have compatible names, must have
+ *                compatible offsets within the datatype and must be in
+ *                the order of "real" part -> "imaginary" part, such that
+ *                the compound datatype matches the following representation:
+ *
+ *                H5T_COMPOUND {
+ *                    <float_type> "r(e)(a)(l)";                OFFSET 0
+ *                    <float_type> "i(m)(a)(g)(i)(n)(a)(r)(y)"; OFFSET SIZEOF("r(e)(a)(l)")
+ *                }
+ *
+ *                where "r(e)(a)(l)" means the field may be named any
+ *                substring of "real", such as "r", or "re" and
+ *                "i(m)(a)(g)(i)(n)(a)(r)(y)" means the field may be named
+ *                any substring of "imaginary", such as "im" or "imag".
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5T__conv_complex_compat(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata,
```

Comment:
This function is just a no-op conversion when converting between complex number datatypes and other datatypes that are in-memory compatible. This provides efficient conversions from h5py-written data in compound datatype format to in-memory complex number format.